acados/ocp_nlp/ocp_nlp_common.c

```diff
@@ -2115,11 +2242,208 @@ void ocp_nlp_embed_initial_value(ocp_nlp_config *config, ocp_nlp_dims *dims,
 }
 
 
+double ocp_nlp_compute_merit_gradient(ocp_nlp_config *config, ocp_nlp_dims *dims,
+                                  ocp_nlp_in *in, ocp_nlp_out *out, ocp_nlp_opts *opts,
+                                  ocp_nlp_memory *mem, ocp_nlp_workspace *work)
+{
+    /* computes merit function gradient at iterate: out -- using already evaluated gradients of submodules
+       with weights: work->weight_merit_fun */
+    int i, j;
+
+    int N = dims->N;
+    int *nv = dims->nv;
+    int *nx = dims->nx;
+    int *nu = dims->nu;
+    int *ni = dims->ni;
+
+    double merit_grad = 0.0;
+    double weight;
+
+    // NOTE: step is in: mem->qp_out->ux
+    struct blasfeo_dvec *tmp_vec; // size nv
+    struct blasfeo_dvec tmp_vec_nxu = work->tmp_nxu;  // size nx
+    struct blasfeo_dvec dxnext_dy = work->dxnext_dy;  // size nx
+
+    // cost
+    for (i=0; i<=N; i++)
+    {
+        tmp_vec = config->cost[i]->memory_get_grad_ptr(mem->cost[i]);
+        merit_grad += blasfeo_ddot(nv[i], tmp_vec, 0, mem->qp_out->ux + i, 0);
+    }
+    double merit_grad_cost = merit_grad;
+
+    /* dynamics */
+    double merit_grad_dyn = 0.0;
+    for (i=0; i<N; i++)
+    {
+        // get shooting node gap x_next(x_n, u_n) - x_{n+1};
+        tmp_vec = config->dynamics[i]->memory_get_fun_ptr(mem->dynamics[i]);
+
+        /* compute directional derivative of xnext with direction y -> tmp_vec_nxu */
+        blasfeo_dvecse(nx[i], 0.0, &dxnext_dy, 0);
```

Comment:
I don't really understand these two loops.
Aren't you just trying to compute what dgemv_t would do at once?
And in any case, the index of the matrix BAbt is missing (i.e. `mem->qp_in->BAbt+i`), and the size of the operation is wrong, since it should be the next value of nx (i.e. `nx[i+1]`), at least if I got it correctly.

Reply:
You're right!
It is fixed in https://github.com/acados/acados/pull/772/commits/880ab5790b574d9f177c64a45e241b22cb804c01